### PR TITLE
Switch Service request to a generic

### PR DESCRIPTION
**Pull request by hawkw**
Depends on tower-rs/tower#109.

This branch updates `tower-h2` to work with the change to
`tower-service` that makes services generic over the request type.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: src/client/connect.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ use std::error::Error;
 use std::fmt;
 use std::marker::PhantomData;
 
-/// Establishes an H2 client connection.
+/// Establishes an H2 client connection.\
```
Comment (jonhoo) [inline src/client/connect.rs]: Why this change?

File: src/server/mod.rs [hunk 2]
```diff
@@ -120,7 +120,7 @@ enum PollMain {
 // ===== impl Server =====
 
 impl<S, E, B> Server<S, E, B>
-where S: NewService<Request = Request<RecvBody>, Response = Response<B>>,
+where S: NewService<Request<RecvBody>, Response = Response<B>>,
```
Comment (jonhoo) [inline src/server/mod.rs]: I know that this is semantically the change from https://github.com/tower-rs/tower/pull/109, but it does make me wonder why we treat `Request` as special, but not `Response`. I guess in some sense, `Request` is the choice of the user, whereas the `Response` is a property of the `Service`, but it still just _feels_ weird.

File: src/service.rs [hunk 3]
```diff
@@ -10,9 +10,7 @@ use futures::{Future, Poll};
 /// This is not intended to be implemented directly. Instead, it is a trait
 /// alias of sorts. Implements the `tower_service::Service` trait using
 /// `http::Request` and `http::Response` types.
-pub trait HttpService: ::sealed::Sealed {
-    /// Request payload.
-    type RequestBody: Body;
+pub trait HttpService<RequestBody>: ::sealed::GenericSealed1<RequestBody> {
```
Comment (jonhoo) [inline src/service.rs]: Hmm, we don't *have* to also change this trait to match the new `Service<Request>` style; we could just make the change internally. Is there a particularly good reason for why `RequestBody` should be turned into a generic type argument on `HttpService`?

Comment (jonhoo) [reply src/service.rs]: Also, this no longer requires that `RequestBody: Body`. Is that intentional?

File: src/client/connect.rs (hunk 1 above)
Comment (hawkw) [reply src/client/connect.rs]: That looks like a typo on my part, thanks for catching it!

File: src/service.rs (hunk 3 above)
Comment (hawkw) [reply src/service.rs]: @jonhoo 
> Hmm, we don't *have* to also change this trait to match the new `Service<Request>` style 

I believe we _do_ need to change this trait to be generic. Otherwise, in the blanket impl of `HttpService` for `T: Service<...>`, the body type parameter is unbounded.

This code:
```rust
impl<T, B1, B2> HttpService for T
where T: Service<Request<B1>,
                Response = Response<B2>>,
      B1: Body,
      B2: Body,
{
    type RequestBody = B1;
    type ResponseBody = B2;
    // ...
}
```
fails to compile with this error:
```
error[E0207]: the type parameter `B1` is not constrained by the impl trait,self type, or predicates
  --> src/service.rs:66:9
   |
66 | impl<T, B1, B2> HttpService for T
   |         ^^ unconstrained type parameter
```

> Also, this no longer requires that `RequestBody: Body`. Is that intentional?

No, that's an oversight on my part. Thanks for catching it!

Comment (hawkw) [reply src/service.rs]: Put the `RequestBody: Body` bound back in d27d22a619c8f204caa6c3bef935d99267281597.

File: src/server/mod.rs (hunk 2 above)
Comment (hawkw) [reply src/server/mod.rs]: Yeah, I agree that this feels strange. However, I _think_ that semantically it's the right thing --- you'd expect a `Service` to be able to accept a set of multiple (probably bounded) request types, but respond with a fixed response type that's specified at the `Service` implementation. It _looks_ weird, but semantically feels more natural IMO.

File: src/service.rs (hunk 3 above)
Comment (jonhoo) [reply src/service.rs]: Hmm, interesting. I guess it does have to be that way, it just doesn't immediately make sense to me why that doesn't compile. Oh well, I guess we're changing `HttpService` too then :/

Comment (hawkw) [reply src/service.rs]: Apparently (https://github.com/tower-rs/tower/pull/109#discussion_r228031054) the type system may support this post-chalk.
